core/src/com/unciv/ui/worldscreen/unit/UnitActions.kt

```diff
@@ -299,96 +303,86 @@ object UnitActions {
         }
     }
 
-    private fun addUnitUpgradeAction(unit: MapUnit, actionList: ArrayList<UnitAction>) {
-        val upgradeAction = getUpgradeAction(unit)
+    private fun addUnitUpgradeAction(
+        unit: MapUnit,
+        actionList: ArrayList<UnitAction>,
+        maxSteps: Int = Int.MAX_VALUE
+    ) {
+        val upgradeAction = getUpgradeAction(unit, maxSteps)
         if (upgradeAction != null) actionList += upgradeAction
     }
 
-    fun getUpgradeAction(unit: MapUnit): UnitAction? {
-        val tile = unit.currentTile
+    /**  Common implementation for [getUpgradeAction], [getFreeUpgradeAction] and [getAncientRuinsUpgradeAction] */
+    private fun getUpgradeAction(
+        unit: MapUnit,
+        maxSteps: Int,
+        isFree: Boolean,
+        isSpecial: Boolean
+    ): UnitAction? {
         if (unit.baseUnit().upgradesTo == null) return null
-        if (!unit.canUpgrade()) return null
-        if (tile.getOwner() != unit.civInfo) return null
-        
-        val upgradedUnit = unit.getUnitToUpgradeTo()
-        val goldCostOfUpgrade = unit.getCostOfUpgrade()
+        val unitTile = unit.getTile()
+        val civInfo = unit.civInfo
+        if (!isFree && unitTile.getOwner() != civInfo) return null
+
+        val upgradesTo = unit.baseUnit().upgradesTo
+        val specialUpgradesTo = unit.baseUnit().specialUpgradesTo
+        val upgradedUnit = when {
+            isSpecial && specialUpgradesTo != null -> civInfo.getEquivalentUnit (specialUpgradesTo)
+            isFree && upgradesTo != null -> civInfo.getEquivalentUnit(upgradesTo)  // getUnitToUpgradeTo can't ignore tech
+            else -> unit.getUnitToUpgradeTo(maxSteps)
+        }
+        if (!unit.canUpgrade(unitToUpgradeTo = upgradedUnit, ignoreRequirements = isFree, ignoreResources = true))
+            return null
 
-        return UnitAction(UnitActionType.Upgrade,
-            title = "Upgrade to [${upgradedUnit.name}] ([$goldCostOfUpgrade] gold)",
-            action = {
-                val unitTile = unit.getTile()
-                unit.destroy()
-                val newUnit = unit.civInfo.placeUnitNearTile(unitTile.position, upgradedUnit.name)
+        // Check _new_ resource requirements (display only - yes even for free or special upgrades)
+        // Using Counter to aggregate is a bit exaggerated, but - respect the mad modder.
+        val delta = Counter<String>()
```

Comment:
It doesn't look like you calculate the actual delta, though? I mean, you get the resource requirements of the upgraded unit, but don't remove the resource requirements of the current unit? 
Also, how about renaming to "newResourcesRequired" and "deltaStr" to "newResourceRequirementsString", deltaStr is such an uninformative name

Reply:
Ah yes here it gets clearer what I mean - I removed the wrong unwanted stuff from the version from the screenshots - where you would get "-1 Iron, +1 Aluminum"...